witness/src/Dscp/Snowdrop/Expanders.hs

```diff
@@ -48,73 +50,124 @@ getByGTx ::
+    txType = StateTxType $ getId (Proxy @TxIds) PublicationTxTypeId
+
+seqExpandersPublicationTx :: SeqExpanders Exceptions Ids Proofs Values ctx PublicationTxWitnessed
+seqExpandersPublicationTx =
+    SeqExpanders $ one $ Expander
+        (Set.fromList [publicationOfPrefix])
+        (Set.fromList [accountPrefix, publicationHeadPrefix, publicationOfPrefix])
+        $ \PublicationTxWitnessed
+            { ptwTx = PublicationTx
+                { ptAuthor
+                , ptPrevBlock
+                , ptBlock
+                }
+            } -> do
+                maybePub <- queryOne (PublicationsOf ptAuthor)
+
+                let change = if isJust maybePub then Upd else New
+
+                return $ mkDiffCS $ Map.fromList
+                  [ PublicationsOf  ptAuthor ==> change (LastPublication ptBlock)
+                  , PublicationHead ptBlock  ==> New    (PublicationNext ptPrevBlock)
+                  ]
+
+----------------------------------------------------------------------------
+-- Money tx
+----------------------------------------------------------------------------
 
 seqExpandersGTx ::
        GTxWitnessed -> SeqExpanders Exceptions Ids Proofs Values ctx GTxWitnessed
 seqExpandersGTx (GMoneyTxWitnessed _) =
     contramap (\(GMoneyTxWitnessed tx) -> tx) seqExpandersBalanceTx
-
-----------------------------------------------------------------------------
--- Money tx
-----------------------------------------------------------------------------
+seqExpandersGTx (GPublicationTxWitnessed _) =
+    contramap (\(GPublicationTxWitnessed ptx) -> ptx) seqExpandersPublicationTx
 
 toProofBalanceTx :: TxWitnessed -> (StateTxType, Proofs)
 toProofBalanceTx (TxWitnessed tx (TxWitness {..})) =
     (txType, AddressTxWitness $ WithSignature {..})
   where
-    txType = StateTxType $ getId (Proxy @TxIds) AccountTxId
+    txType = StateTxType $ getId (Proxy @TxIds) AccountTxTypeId
     wsSignature = txwSig
     wsPublicKey = txwPk
-    wsBody = (txId tx, wsPublicKey)
+    wsBody = (txId tx, wsPublicKey, ())
 
 seqExpandersBalanceTx :: SeqExpanders Exceptions Ids Proofs Values ctx TxWitnessed
 seqExpandersBalanceTx =
     SeqExpanders $ one $ Expander inP outP $ \TxWitnessed{..} -> do
         let outputs = txOuts twTx
         -- check for output duplicates
         let uniqOutAddrs = ordNub $ map txOutAddr outputs
+
         when (length outputs > 1 && length uniqOutAddrs /= length outputs) $
             throwLocalError MTxDuplicateOutputs
 
         -- Account we transfer money from
-        let inAddr = tiaAddr $ txInAcc twTx
-        let getCurAcc (fromAddr :: Address) = do
-                (acc :: Maybe Account) <- queryOne (AccountId fromAddr)
-                pure acc
-        let outSame = find ((== inAddr) . txOutAddr) (txOuts twTx)
-        let outOther = maybe outputs (`L.delete` outputs) outSame
+        let inAddr    = tiaAddr $ txInAcc twTx
+        let getCurAcc = queryOne @_ @(AccountId Address) @_ @Account . AccountId
```

Comment:
Please refrain from making subjective changes if it's possible.

Reply:
Was your intent different from constraining the type of `queryOne` in this line?